<a href="https://colab.research.google.com/github/VictoryAzuka/FlyRank_ML_Updated/blob/main/work/notebooks/w01_research_question_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/VictoryAzuka/FlyRank_ML_Updated/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

### Downloading Skill Files

The following cell will download `skills/README.md` and `framing-ml-problems/SKILL.md` from the GitHub repository, making them available for the notebook.

In [10]:
import os
import requests

base_github_url = 'https://raw.githubusercontent.com/VictoryAzuka/FlyRank_ML_Updated/main/'

# --- Download skills/README.md ---
skill_readme_path = 'skills/README.md'
skill_readme_url = base_github_url + skill_readme_path

# Ensure the directory exists
os.makedirs(os.path.dirname(skill_readme_path), exist_ok=True)

print(f"Downloading {skill_readme_path} from {skill_readme_url}...")
response_skills_readme = requests.get(skill_readme_url)
response_skills_readme.raise_for_status() # Raise an exception for bad status codes

# Save the file locally
with open(skill_readme_path, 'wb') as f:
    f.write(response_skills_readme.content)
print(f"Successfully downloaded and saved to {skill_readme_path}")

print("\n" + "="*50 + "\n") # Separator for clarity

# --- Download framing-ml-problems/SKILL.md ---
framing_skill_path = 'skills/framing-ml-problems/SKILL.md' # Corrected path
framing_skill_url = base_github_url + framing_skill_path

# Ensure the directory exists
os.makedirs(os.path.dirname(framing_skill_path), exist_ok=True)

print(f"Downloading {framing_skill_path} from {framing_skill_url}...")
response_framing_skill = requests.get(framing_skill_url)
response_framing_skill.raise_for_status() # Raise an exception for bad status codes

# Save the file locally
with open(framing_skill_path, 'wb') as f:
    f.write(response_framing_skill.content)
print(f"Successfully downloaded and saved to {framing_skill_path}")

Successfully downloaded and saved to skills/README.md


Successfully downloaded and saved to skills/framing-ml-problems/SKILL.md


In [11]:
import os

# Read skills/README.md
file_path_skills = 'skills/README.md'
if os.path.exists(file_path_skills):
    print(f"--- Content of {file_path_skills} ---")
    with open(file_path_skills, 'r') as f:
        content_skills = f.read()
    print(content_skills)
else:
    print(f"Error: File '{file_path_skills}' not found. Please ensure it is in the correct directory.")

print("\n" + "="*50 + "\n") # Separator for clarity

# Read framing-ml-problems/SKILL.md
file_path_framing = 'skills/framing-ml-problems/SKILL.md' # Corrected path
if os.path.exists(file_path_framing):
    print(f"--- Content of {file_path_framing} ---")
    with open(file_path_framing, 'r') as f:
        content_framing = f.read()
    print(content_framing)
else:
    print(f"Error: File '{file_path_framing}' not found. Please ensure it is in the correct directory.")

--- Content of skills/README.md ---
# Skills — the router

This folder is a small library of **skills**: focused instruction files your AI assistant loads
one at a time. One skill per task keeps the assistant sharp — its context window is small, and
filling it with everything makes it worse at the one thing you need.

**How to use it (repo-reading agents — Claude Code, Cursor, Codex):** they find this file
automatically via `AGENTS.md` / `CLAUDE.md`. Just tell your assistant which task you're doing.

**Using a chat-only assistant (ChatGPT / Gemini in a browser)?** Open the skill file on GitHub,
copy its whole content, and paste it into your chat before asking for help. That's it.

## The table — find your task, load ONE skill

| Your task | Load this skill | Also load for data work |
|---|---|---|
| Any task — how to work with your assistant at all | `directing-your-ai-assistant/SKILL.md` | — |
| Pick a lane, frame your question (ML-02, ML-03) | `framing-ml-problems/SKILL.md` | `flyran

# 1. My lane (or freestyle) and why

### My lane is Supervised (CTR / Engagement Opportunity Scoring). This is because the models learn from already labelled data with known target outcomes.

In [12]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


# 2. The question: decision, action, cost of a wrong call

### My work helps the editor to know which page(s) to improve.

### The Editor is the one who acts on the output. He/she tracks key CTR engagement and performance reporting to verify whether the fixes improved engagements.


 ### The cost of a wrong recommendation is failing to flag a page that is performing poorly, leading to lost traffic, lower search visibility, and reduced user engagement.

In [13]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [14]:
import os
import requests

# Define the directory and file path
data_dir = 'data/raw'
csv_filename = 'content_refresh_anonymized.csv'
csv_filepath = os.path.join(data_dir, csv_filename)

# Create the directory if it doesn't exist
os.makedirs(data_dir, exist_ok=True)

# Define the URL for the raw CSV file on GitHub
github_csv_url = 'https://raw.githubusercontent.com/VictoryAzuka/FlyRank_ML_Updated/main/data/raw/content_refresh_anonymized.csv'

print(f"Downloading {csv_filename} from {github_csv_url}...")
response = requests.get(github_csv_url)
response.raise_for_status() # Raise an exception for bad status codes

# Save the file locally
with open(csv_filepath, 'wb') as f:
    f.write(response.content)

print(f"Successfully downloaded and saved to {csv_filepath}")

Successfully downloaded and saved to data/raw/content_refresh_anonymized.csv


In [15]:
import pandas as pd

csv_path = 'data/raw/content_refresh_anonymized.csv'
df = pd.read_csv(csv_path)

print("First 5 rows of the dataset:")
display(df.head())

First 5 rows of the dataset:


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [16]:
print("\nDescriptive statistics for key columns:")
display(df[['ctr', 'engagement_rate', 'scroll_rate', 'trend_pct', 'avg_position']].describe())


Descriptive statistics for key columns:


,ctr,engagement_rate,scroll_rate,trend_pct,avg_position
count,30000.000000,30000.000000,29875.000000,26612.000000,30000.00000
mean,0.510733,2.534520,18.212921,-4.785969,16.34238
std,3.279162,8.310096,29.472768,473.861780,15.21679
min,0.000000,0.000000,0.000000,-100.000000,0.00000
25%,0.000000,0.000000,0.000000,-62.600000,6.20000
50%,0.070000,0.000000,5.000000,-33.500000,10.80000
75%,0.290000,1.350000,23.530000,0.000000,22.30000
max,100.000000,100.000000,300.000000,44900.000000,245.00000


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

## Observed
- Traffic decay is widespread across the dataset, with 50\% of pages experiencing a traffic loss of 33.5% or worse trend_pct median is -33.5\%, mean is -4.79%.

- A substantial portion of content sits in striking distance of top rankings, with 50% of pages holding an average rank of position 10.8 or better (avg_position median is 10.8).

- The raw metrics contain extreme upper-bound outliers and data noise, including maximum click-through rates and engagement rates recorded at 100% and scroll rates reaching up to 300%.

## Directional
- Relative scoring rank orders pages by their expected potential return on editing effort, indicating which decaying pages show higher opportunity relative to others.

- Trend indicators (trend_pct median of -33.5%) identify trajectories of decaying traffic versus stable or growing content across individual page units where 1 text row = 1 text page.

- High-opportunity signals (such as striking-distance average positions with a median rank of 10.8) point toward pages that are closer to page-one visibility, prioritizing them over low-impact or deep-page content.

## Decision Support

# What it will say:

- PRIORITY RANKING: Ranks pages by content refresh opportunity based on historical performance signals (such as striking-distance position ranks and trend decay rates).

RESOURCE ALLOCATION: Identifies which top subset of decaying content offers the highest potential return on editing effort.

TRIAGE & FILTERING: Flags high-volume pages experiencing traffic decay so editors can prioritize re-optimizing them over low-impact content.

# What it will NEVER say:

CAUSAL PROOF: Does not guarantee that updating a page will directly cause Google rankings or traffic to increase.

PREDICTING GOOGLE: Does not model or predict Google's proprietary search engine ranking algorithm.

AUTOMATED ACTION: Does not automatically rewrite content or declare pages broken without human editorial review and approval.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.